dsn_test.go

```diff
@@ -1639,3 +1640,58 @@ func TestExtractAccountName(t *testing.T) {
 		})
 	}
 }
+
+func TestUrlDecodeIfNeeded(t *testing.T) {
+	testcases := map[string]string{
+		"query_tag":             "query_tag",
+		"%24my_custom_variable": "$my_custom_variable",
+	}
+	for param, expected := range testcases {
+		t.Run(param, func(t *testing.T) {
+			decodedParam := urlDecodeIfNeeded(param)
+			if decodedParam != expected {
+				t.Fatalf("urlDecodeIfNeeded returned unexpected response (%v), should be %v", decodedParam, expected)
+			}
+		})
+	}
+}
+
+func TestUrlDecodeIfNeededE2E(t *testing.T) {
+	customVarName := "CUSTOM_VARIABLE"
+	customVarValue := "test"
+	myQueryTag := "mytag"
+
+	cfg := &Config{
+		Account:  os.Getenv("SNOWFLAKE_TEST_ACCOUNT"),
+		User:     os.Getenv("SNOWFLAKE_TEST_USER"),
+		Password: os.Getenv("SNOWFLAKE_TEST_PASSWORD"),
+		Params:   map[string]*string{"$" + customVarName: &customVarValue, "query_tag": &myQueryTag},
+	}
+	mydsn, err := DSN(cfg)
+	if err != nil {
+		t.Fatalf("TestUrlDecodeIfNeededE2E failed to create DSN from Config: %v, err: %v", cfg, err)
+	}
+	db, err := sql.Open("snowflake", mydsn)
+	if err != nil {
+		t.Fatalf("TestUrlDecodeIfNeededE2E failed to connect. %v, err: %v", mydsn, err)
+	}
+	defer db.Close()
+	query := "SHOW /* gosnowflake TestUrlDecodeIfNeededE2E */ VARIABLES;"
+	rows, err := db.Query(query)
+	if err != nil {
+		t.Fatalf("TestUrlDecodeIfNeededE2E failed to run SHOW VARIABLES query, error: %v", err)
+	}
+	defer rows.Close()
+	var v1, v2, v3, v4, v5, v6, v7 any
+	for rows.Next() {
```

Comment:
I think it's better to assert on this (`assertTrueF(t, rows.Next())`). If no rows are returned, the test will end up correctly, but skipping this section at all.

Reply:
the test expects to read the same session variable back from the rows returned, so if no rows are returned that's a bad thing and test should fail (existing assertion `assertDeepEqualE` should also cover that already)

Reply:
(got this now what you meant, after the internal discussion , sent a commit accordingly)